I Can Hear a Rainbow: scripts/globals to scripts/quests

wrenffxi opened the pull request:
<!-- place 'x' mark between square [] brackets to affirm: -->
**_I affirm:_**
- [x] that I agree to Project Topaz's [Limited Contributor License Agreement](http://project-topaz.com/blob/release/CONTRIBUTOR_AGREEMENT.md), as written on this date
- [x] that I've _tested my code_ since the last commit in the PR, and will test after any later commits

* remove from scripts/globals, and rewrite in scripts/quests
* make the color data more easily readable
* localVar the variable used in eventUpdate, which removes the need for onEventFinish
* clear the correct charVar on quest completion
* rainbow controlled by localVar on rainbow NPC, which removes need for server variable

update: per feedback,

* ignore weather set by SCH<sup>1</sup>
* use set constructor for color data
* swap order of quest check vs ruby check

<sup>1</sup>Two issues had to be solved here.  First, there wasn't a way to get weather from zone into LUA.  To resolve this, I added a zone:getWeather() binding.  Second, during onZoneIn, the players's zone is not yet set, causing player:getZone() to return null.  To resolve this, I added an optional bool isZoning parameter to player:getZone(), which will instead return the player's destination zone.

wrenffxi committed:
i can hear a rainbow

TeoTwawki commented:
thankyouwrenthatscriptnamealwaysbotheredme

zach2good reviewed (approved):
I was looking at this the other day hoping someone would clean it up, and here it is! Nice work!

ibm2431 commented on scripts/quests/i_can_hear_a_rainbow.lua:
```diff
@@ -0,0 +1,158 @@
+            [tpz.zone.YUHTUNGA_JUNGLE] = true,
+            [tpz.zone.YHOATOR_JUNGLE] = true,
+            [tpz.zone.WESTERN_ALTEPA_DESERT] = true,
+        },
```
We have a [set constructor](https://github.com/project-topaz/topaz/blob/release/scripts/globals/common.lua#L35-L45) which will do this for you from a simple table:
```lua
zones = set{
    tpz.zone.VALKURM_DUNES,
    tpz.zone.ROLANBERRY_FIELDS,
    tpz.zone.CAPE_TERIGGAN,
   -- etc
}

ibm2431 commented on scripts/quests/i_can_hear_a_rainbow.lua:
```diff
@@ -0,0 +1,158 @@
+-------------------------------------------------
+
+quests.i_can_hear_a_rainbow.onZoneIn = function(player)
+    if player:hasItem(1125, 0) and player:getQuestStatus(WINDURST, tpz.quest.id.windurst.I_CAN_HEAR_A_RAINBOW) == QUEST_ACCEPTED then
```
Maybe swap order of quest status check and `hasItem`. HasItem iterates over everything in the bag, whereas quest status is a quick return. Checking for status first means we only ping inventory if player is on the quest.

ibm2431 commented on scripts/quests/i_can_hear_a_rainbow.lua:
```diff
@@ -0,0 +1,158 @@
+        local trigger = false
+
+        -- get data for basic weather (e.g. downgrade GALES to WIND)
+        local weather = player:getWeather()
```
`entity:getWeather()` will return SCH weather too; are those lost on zoning? Does the effect loss include logging?

TeoTwawki replied:
use zone entities weather instead. sch 'weather' doesn't work for smn quest

wrenffxi replied:
I don't currently see a way to get zone's weather into LUA, so will have to either add a getWeather binding to lua_zone, and/or add an optional ignoreScholar bool parameter to player:getWeather.  Any preference?

TeoTwawki replied:
when the object type isn't a player or mob, it assumed we want zone weather:
```lua
inline int32 CLuaBaseEntity::getWeather(lua_State *L)
{
    TPZ_DEBUG_BREAK_IF(m_PBaseEntity == nullptr);

    WEATHER weather = WEATHER_NONE;
    if (m_PBaseEntity->objtype & TYPE_PC || m_PBaseEntity->objtype & TYPE_MOB)
        weather = battleutils::GetWeather((CBattleEntity*)m_PBaseEntity, false);
    else
        weather = zoneutils::GetZone(m_PBaseEntity->getZone())->GetWeather();

    lua_pushinteger(L, weather);

    return 1;
}
```
so you can get the zone (from the player if need be) and then
`zone:getWeather()`
works

TeoTwawki replied:
`player:getZone():getWeather()`

for example

TeoTwawki replied:
Sorry I see what you are saying about it not being in lua_zone...I thought we already had this functionality in use elsewhere already.

Quick search says I am wrong

wrenffxi committed:
add binding zone:getWeather()
add optional isZoning param to player:getZone()
use player:getZone(isZoning):getWeather() in ichar script
use set constructor in ichar script
swap order of quest check and hasItem

wrenffxi commented:
Updated PR per feedback.  Description lists changes.

ibm2431 reviewed (approved):
Thanks, Wren~
